lib/origen_testers/igxl_based_tester/ultraflex.rb

```diff
@@ -498,11 +496,213 @@ def store_next_cycle(*pins)
+      # issues the 'Start <waveform>' microcode
+      # Required arguments:
+      #                     pins
+      # Optional arguments:
+      #                     signal-name  (DigSrc waveform)
+      #                     starte   (true or false, defaults to false)
+      # Note restrestions from IG-XL help around use of starte opcode spacing to end of the current segment.
+      def digsrc_start(*pins)
+        options = pins.last.is_a?(Hash) ? pins.pop : {}
+        options = {
+          signal_name: '',
+          starte:      false
+        }.merge(options)
+
+        pins = pins.flatten.compact
+        options[:starte] ? start_syntax = 'Starte' : start_syntax = 'Start'
+        microcode "(#{format_multiple_instrument_pins(pins)}:DigSrc = #{start_syntax} #{options[:signal_name]})"
+
+        if options[:dssc_mode] == :single
+          $tester.cycle(repeat: 145) # minimum of 144 cycles, adding 1 for safey measures
+        elsif options[:dssc_mode] == :dual
+          $tester.cycle(repeat: 289) # minimum of 288 cycles, adding 1 for safety measures
+        else
+          $tester.cycle(repeat: 577) # minimum of 577 cycles, adding 1 for safety measures
+        end
+      end
+
+      # Call this method at the end of each digsrc overlay operation to clear the pattern
+      # memory pipeline, so that the pattern is ready to do the next digsrc overlay operation.
+      # Required arguments:
+      #                     pins
+      def digsrc_stop(*pins)
+        pins = pins.flatten.compact
+        microcode "(#{format_multiple_instrument_pins(pins)}:DigSrc = Stop)"
+      end
+
+      # Call this method to insert the digsrc SEND overlay microcode on the previous vector and change the pin state.
+      def digsrc_send(*pins)
+        dssc(pins, dssc_microcode: 'DigSrc = Send', pin_state: 'drive_mem')
+      end
+
+      # Call this method to insert the digsrc SEND overlay microcode on the next vector
+      def digsrc_send_next_cycle(*pins)
+        dssc_next_cycle(pins, dssc_microcode: 'DigSrc = Send', pin_state: 'drive_mem')
+      end
+      alias_method :digsrc_send!, :digsrc_send_next_cycle
+
+      # Call this method to insert the digcap TRIG microcode on the previous vector
+      def digcap_trig(*pins)
+        dssc(pins, dssc_microcode: 'DigCap = Trig')
+      end
+
+      # Call this method to insert the digcap TRIG microcode on the next vector
+      def digcap_trig_next_cycle(*pins)
+        dssc_next_cycle(pins, dssc_microcode: 'DigCap = Trig')
+      end
+      alias_method :digcap_trig!, :digcap_trig_next_cycle
+
+      # Call this method to insert the digcap STORE microcode on the previous vector
+      def digcap_store(*pins)
+        options = pins.last.is_a?(Hash) ? pins.pop : {}
+        options = {
+          offset:         0,
+          dssc_microcode: 'DigCap = Store'
+        }.merge(options)
+        dssc(pins, options)
+      end
```

Comment:
Store could also be used as-is, provided the instruments statement enables 2 DigCap options:  'auto_trig_enable' and 'store_stv'.

 `instruments = {
    TDO : digcap 1 : data_type=long : auto_trig_enable : store_stv_enable : receive_data=logic;
  }`

With 'auto_trig_enable', you don't need the 'DigCap = Trig' opcode.  This mode does have some differences in error handling.

With 'store_stv', the 'stv' opcode is used to capture data.  All capture instruments in the pattern must use this mode, so it may not be useful in cases where a pattern uses multiple instruments (say on both tdo and portA).

This is a case where having the instruments statement in a separate header file (pulled in with pre-processor) is helpful - allows to easily change between HRAM captures and DSSC just by re-compiling.